### ISSUE TITLE
Orleans.Runtime.Catalog uses exceptions in non exceptional cases

### ISSUE THREAD

**Issue by philippecp** [dotnet/orleans#1607]
I have the following scenario: 

- A call activates multiple stateless worker grains concurrently

- Each stateless worker activates the same stateful grain on activation

In this scenario, I have noticed the first time I make this call after silo startup takes significantly longer than following calls. I obviously suspected this was because the grain were being instantiated and loaded from storage. What I did not expect was to see hundreds of exceptions originating from the catalog in my intellitrace history. The two culprits are DuplicateActivationException and NonExistantActivationException. Given that Orleans seems to recover just fine from those exceptions, I would argue that those should not be handled by exceptions but by regular logic. Furthermore, since .Net can only handle a few hundred exceptions per second, using exceptions for "normal" behaviour is known to negatively affect performance.

**Comment by gabikliot** [comment]
I agree with that. Exceptions were used in those cases mostly for historical reasons, it used to be spread across multiple components, but then the code changed....
I think it should be pretty easy to replace with return values that indicate errors.

**Comment by dVakulen** [comment]
In this particular case it indeed will be easy. I'll take a look at it.

**Comment by dVakulen** [comment]
I've made rough implementation of this, and it's already more lines of code than it was before, while not preserving all the guarantees (notifying system grain creation system about fails will add even more lines, which woun't help readability). So it will be big PR with lots of changes in the core, which will require thorough attention from maintainers thus taking their time. On the other hand I've ran benchmarks to measure perf impact of exception throwing and on the  laptop CPU with 2.5 Ghz frequency .Net 4.5 can handle 90 thousands of exception per seconds per core; throughput doesn't drop much with increase of the depth of the call stack. So if the missed activation calls are really the issue of the slow first silo's calls - this area should be revisited more deeply than just removing exceptions throwing. And under such revisit exceptions could be removed awell. 

Abovementioned implementation can be seen at the following lcoation: 
https://github.com/dotnet/orleans/compare/master...dVakulen:Remove_duplicate_activation_exception?expand=1#diff-bcfe9efaf8ce10fd652fca0dd5a4d116L24

**Comment by philippecp** [comment]
Although the raw cost for throwing exceptions isn't that high, it does have additional overhead because exceptions often get logged which usually involves doing IO of some kind.  Yes .Net can handle 90k exceptions per second, but I'm sure it can handle at least a couple orders of magnitude more function calls. Furthermore, exceptions significantly slow down debugging due to the more detailed stack trace as well as exceptions being tracked by intellitrace.

That being said, I agree that the primary reason for this change should not be performance. The real cost is that when investigating an actual issue these specific exceptions generate a lot of noise. Furthermore it's certainly considered bad practice to use exceptions for normal execution flow [as published here](https://blogs.msdn.microsoft.com/kcwalina/2005/03/16/design-guidelines-update-exception-throwing/).

Whether or not the benefits of doing this outweigh the risks is up to the maintainers but I would hope Orleans has enough unit tests to do these kinds of improvements with minimal risks rather than fall into the big ball of mud.

**Comment by sergeybykov** [comment]
The key factor to me is maintainability of the code. Additional complexity is justified when there are significant performance gains, e.g. like in the messaging/buffer pool case.

If somebody proposes a PR where the code isn't much more complex than it is today, we'll gladly review and likely merge it. It sounds like @dVakulen himself isn't sure if the approach he tried to take will end up worth it, unless I misunderstood.

I agree that there's a bit of controversy here that exceptions are used for something that is part of the normal case protocol. This does cause some confusion because people instinctively read such logs as signs of potential problems. Coincidently, I had a conversation about such a case just this morning. So improvements here are more than welcome.

**Comment by dVakulen** [comment]
Ok, will try to finish the implementation.

**Comment by dVakulen** [comment]
@sergeybykov  I haven't forgotten about this one, just trying to implement fix to the root of the problem: in the NightlyLoad test almost every forth grain activation becomes duplicated, with following request rerouting. Solution is based on assumption that 95% of the duplicate - causing requests are issued from the same client\silo.

**Comment by sergeybykov** [comment]
I'm afraid this is where the load test may be misleading. It starts in parallel a number of load generators that all round-robin over the same set of grains. The initial burst of activations by the uncoordinated load generators causes the inefficiency of duplicate attempts to activate same grains and eventual forwarding of messages to the 'winner' activations

However, in a real production case it is presumably much less likely that requests to the same grains will nearly simultaneously arrive from multiple clients via different gateways. And that's what we see in production - duplicate activation attempts are relatively infrequent.

**Comment by dVakulen** [comment]
Almost 50% of requests causes duplicate activations even when running local setup with concurrent requests to the same grain. That's due to relatively long activation creation process during which additional requests may come in. That's a bit different issue from the one I've mentioned earlier, and may be not a problem on real deployments.

**Comment by dVakulen** [comment]
If removing of the exception still has some priority - I'll try to finish my implementation.

**Comment by sergeybykov** [comment]
>Almost 50% of requests causes duplicate activations even when running local setup with concurrent requests to the same grain. That's due to relatively long activation creation process during which additional requests may come in. That's a bit different issue from the one I've mentioned earlier, and may be not a problem on real deployments too.

In our experience, this is not a problem in most scenarios, as requests to a grain are usually spread over time. And the cost of handling it it's that high - an extra roundtrip between silos.

**Comment by ReubenBond** [comment]
Both of these cases are fixed starting in 4.0.0-preview1 and we have removed `DuplicateActivationException` and `NonExistentActivationException` entirely. The concept of a duplicate activation still exists, but it's handled internally by the activation (they will forward their messages onto the successor). Stateless workers can no longer create duplicate activations by design. The occurrence rate of duplicate activations was also reduced dramatically by redesigning how placement works internally (see #7154). As for `NonExistentActivation`, we removed the concept of 'multiple activation' grains from the guts of Orleans, so the catalog will happily activate a grain as soon as it receives a message for it (previously there was an ordering issue, where a certain flag needed to be set on a message header in order for a new activation to be created by the catalog).
